src/traced/probes/ftrace/tracefs.cc

```diff
@@ -450,21 +450,87 @@ size_t Tracefs::NumberOfCpus() const {
   return num_cpus;
 }
 
+size_t Tracefs::NumberOfOnlineCpus() const {
+  return static_cast<size_t>(sysconf(_SC_NPROCESSORS_ONLN));
+}
+
+std::optional<std::vector<uint32_t>> Tracefs::GetOfflineCpus() const {
+  std::string offline_cpus_str;
+  if (!ReadFile("/sys/devices/system/cpu/offline", &offline_cpus_str)) {
+    PERFETTO_ELOG("Failed to read offline cpus file");
+    return std::nullopt;
+  }
+  offline_cpus_str = base::TrimWhitespace(offline_cpus_str);
+
+  // The offline cpus file contains a list of comma-separated CPU ranges.
+  // Each range is either a single CPU or a range of CPUs, e.g. "0-3,5,7-9".
+  // Source: https://docs.kernel.org/admin-guide/cputopology.html
+  std::vector<uint32_t> offline_cpus;
+  for (base::StringSplitter ss(offline_cpus_str, ','); ss.Next();) {
+    base::StringView offline_cpu_range(ss.cur_token(), ss.cur_token_size());
+    size_t dash_pos = offline_cpu_range.find('-');
+    if (dash_pos == base::StringView::npos) {
+      // Single CPU in the format of "%d".
+      std::optional<uint32_t> cpu = base::StringViewToUInt32(offline_cpu_range);
+      if (cpu.has_value()) {
+        offline_cpus.push_back(cpu.value());
+      } else {
+        PERFETTO_ELOG("Failed to parse single CPU from offline CPU range: %s",
+                      offline_cpu_range.data());
+        return std::nullopt;
+      }
+    } else {
+      // Range of CPUs in the format of "%d-%d".
+      std::optional<uint32_t> start_cpu =
+          base::StringViewToUInt32(offline_cpu_range.substr(0, dash_pos));
+      std::optional<uint32_t> end_cpu =
+          base::StringViewToUInt32(offline_cpu_range.substr(dash_pos + 1));
+      if (start_cpu.has_value() && end_cpu.has_value()) {
+        for (auto cpu = start_cpu.value(); cpu <= end_cpu.value(); ++cpu) {
+          offline_cpus.push_back(cpu);
+        }
+      } else {
+        PERFETTO_ELOG("Failed to parse CPU range from offline CPU range: %s",
+                      offline_cpu_range.data());
+        return std::nullopt;
+      }
+    }
+  }
+  return offline_cpus;
+}
+
 void Tracefs::ClearTrace() {
   std::string path = root_ + "trace";
   PERFETTO_CHECK(ClearFile(path));  // Could not clear.
 
+  const auto total_cpu_count = NumberOfCpus();
+  const auto online_cpu_count = NumberOfOnlineCpus();
+
   // Truncating the trace file leads to tracing_reset_online_cpus being called
-  // in the kernel.
-  //
-  // In case some of the CPUs were not online, their buffer needs to be
-  // cleared manually.
-  //
-  // We cannot use PERFETTO_CHECK as we might get a permission denied error
-  // on Android. The permissions to these files are configured in
-  // platform/framework/native/cmds/atrace/atrace.rc.
-  for (size_t cpu = 0, num_cpus = NumberOfCpus(); cpu < num_cpus; cpu++) {
-    ClearPerCpuTrace(cpu);
+  // in the kernel. So if all cpus are online, no further action needed.
+  if (total_cpu_count == online_cpu_count)
+    return;
+
+  PERFETTO_DLOG(
```

Comment:
I'd promote this to a log. this is realistically never going to be hit, but if it's hit might help explaining why we are taking more time.